crates/typst-docs/src/main.rs

```diff
@@ -0,0 +1,148 @@
+use std::fs;
+use std::path::{Path, PathBuf};
+
+use clap::Parser;
+use regex::Regex;
+use typst::model::Document;
+use typst::visualize::Color;
+use typst_docs::{provide, Html, Resolver};
+use typst_render::render;
+
+#[derive(Debug)]
+struct MyResolver<'a> {
+    assets_dir: &'a Path,
+    verbose: bool,
+    base: &'a str,
+}
+
+impl<'a> Resolver for MyResolver<'a> {
+    fn commits(&self, from: &str, to: &str) -> Vec<typst_docs::Commit> {
+        if self.verbose {
+            eprintln!("commits({from}, {to})");
+        }
+        vec![]
+    }
+
+    fn example(
+        &self,
+        hash: u128,
+        source: Option<Html>,
+        document: &Document,
+    ) -> typst_docs::Html {
+        if self.verbose {
+            eprintln!(
+                "example(0x{hash:x}, {:?} chars, Document)",
+                source.as_ref().map(|s| s.as_str().len())
+            );
+        }
+
+        let frame = &document.pages.first().expect("page 0").frame;
+        let pixmap = render(frame, 2.0, Color::WHITE);
+        let filename = format!("{hash:x}.png");
+        let path = self.assets_dir.join(&filename);
+        fs::create_dir_all(path.parent().expect("parent")).expect("create dir");
+        pixmap.save_png(path.as_path()).expect("save png");
+        let src = format!("{}assets/{filename}", self.base);
+        eprintln!("Generated example image {path:?}");
+
+        if let Some(code) = source {
+            let code_safe = code.as_str();
+            Html::new(format!(
+                r#"<div class="previewed-code"><pre>{code_safe}</pre><div class="preview"><img src="{src}" alt="Preview" /></div></div>"#
+            ))
+        } else {
+            Html::new(format!(
+                r#"<div class="preview"><img src="{src}" alt="Preview" /></div>"#
+            ))
+        }
+    }
+
+    fn image(&self, filename: &str, data: &[u8]) -> String {
+        if self.verbose {
+            eprintln!("image({filename}, {} bytes)", data.len());
+        }
+
+        let path = self.assets_dir.join(filename);
+        fs::create_dir_all(path.parent().expect("parent")).expect("create dir");
+        fs::write(&path, data).expect("write image");
+        eprintln!("Created {} byte image at {path:?}", data.len());
+
+        format!("{}assets/{filename}", self.base)
+    }
+
+    fn link(&self, link: &str) -> Option<String> {
+        if self.verbose {
+            eprintln!("link({link})");
+        }
+        None
+    }
+}
+
+/// Generates the JSON representation of the documentation. This can be used to
+/// generate the HTML yourself. You are encouraged to post-process the generated
+/// JSON to rewrite links and other things to match your site's structure. Be
```

Comment:
I think we can remove this encouragement for post-processing now with the new `base` config.